Add observation for message channels

artembilan opened the pull request:
<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
Add observation for message channels

garyrussell reviewed (changes requested):
Commit comment would have been helpful.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractMessageChannel.java:
```diff
@@ -293,11 +312,45 @@ public boolean send(Message<?> messageArg, long timeout) {
+	}
+
+	private boolean sendWithObservation(Message<?> message, long timeout) {
+		MutableMessage<?> messageToSend = MutableMessage.of(message);
```
See the `MessageSenderContext` implementation.
The `MessageChannel.send()` is really a point in Spring Integration where we produce messages.
According to `PRODUCER` entity of the observation, we have to expose some `Setter` which would carry on values from the current context to downstream consumers - propagation, essentially.
Therefore a `MutableMessage` to be able to modify headers in the trace `Propagator`.
See `ObservationPropagationChannelInterceptorTests`.

Or is your "Why?" about this new `of()` factory method? 😄 

Sure! I can add more info into commit message if we are really on the same page about the solution after this review and discussion.

Thank you for understanding!

P.S. Perhaps I deliberately have left so little info in the PR, therefore you would pay attention for what is going on and we would have a healthy discussion like this 😉

garyrussell replied:
Yes, I missed the `getHeaders().put(...)`; now I see it.

I will be honest in that I still don't see why we need to observe channel sends, it seems much too fine-grained to me.

artembilan replied:
Well, if channel is distributed, then we just lose the producer point of the trace.
It is really up to end-user now what to chose for instrumentation. 
Perhaps they indeed would apply only inbound and outbound channel adapters.
However the current solution with Sleuth is only channel instrumentation 🤷

garyrussell replied:
Ok; just address my other comment and will merge.

artembilan committed:
Add observation for message channels

The `MessageChannel.send()` is, essentially, only the point in Spring Integration where we produce a message
and can emit a `PRODUCER` kind span.

* Implement `IntegrationObservation.PRODUCER` infrastructure based on the `MessageSenderContext`
* Implement an observation emission in the `AbstractMessageChannel` based on the mentioned `IntegrationObservation.PRODUCER`
* Build a `MutableMessage.of(message)` to be able to modify message header in the `MessageSenderContext` via tracer `Propagator`
or other tracing injection instrument
* Document which components are instrumented with an `ObservationRegistry`

garyrussell reviewed (changes requested):
Doc polishing.

artembilan committed:
Fix language in docs

Co-authored-by: Gary Russell <grussell@vmware.com>
